lib/new_relic/agent/stats.rb

```diff
@@ -122,17 +129,35 @@ def ==(other)
       alias_method :apdex_f, :total_exclusive_time
 
       def record_apdex(bucket, apdex_t)
-        case bucket
-        when :apdex_s then @call_count += 1
-        when :apdex_t then @total_call_time += 1
-        when :apdex_f then @total_exclusive_time += 1
+        @lock.synchronize do
+          case bucket
+          when :apdex_s then @call_count += 1
+          when :apdex_t then @total_call_time += 1
+          when :apdex_f then @total_exclusive_time += 1
+          end
+          if apdex_t
+            @min_call_time = apdex_t
+            @max_call_time = apdex_t
+          else
+            ::NewRelic::Agent.logger.warn("Attempted to set apdex_t to #{apdex_t.inspect}, backtrace = #{caller.join("\n")}")
+          end
         end
-        if apdex_t
-          @min_call_time = apdex_t
-          @max_call_time = apdex_t
-        else
-          ::NewRelic::Agent.logger.warn("Attempted to set apdex_t to #{apdex_t.inspect}, backtrace = #{caller.join("\n")}")
+      end
+
+      # Override marshalling methods to exclude @lock from being included in marshalled data
+      def marshal_dump
+        instance_variables.reject { |name| SKIP_MARSHALLING.include?(name) }.inject({}) do |instance_copy, name|
+          instance_copy[name] = instance_variable_get(name)
+          instance_copy
+        end
```

Comment:
Two things here: 1) I think we should use `#each_with_object`  (available since Ruby 1.9) instead of `#inject`, and 2) I think we should simply perform the skipping of names ourselves, as `#reject` will end up looping over every single value a second time otherwise.

Here's a suggested rewrite:

```ruby
def marshal_dump
  instance_variables.each_with_object({}) do |name, instance_copy|
    next if SKIP_MARSHALLING.include?(name)
    instance_copy[name] = instance_variable_get(name)
  end
end
```

Reply:
I do like that a lot better, it looks a lot nicer, I'll get that updated, thank you!